### PR TITLE
[MRG] add `MinHash.kmers_and_hashes(...)` and `sourmash sig kmers`

### LINKED ISSUE #483: Is there a way to go from the hash value to the k-mer sequence?
I.e. any switch that tracks which k-mer sequence is present in the signature?

Thank you.

### LINKED ISSUE #1372: Extracting k-mers from signatures still possible?
Hi!

I can see in the past there's been a script to convert signatures generated by sourmash back to their respective k-mers which was located in utils directory, but in the current github utils directory the 'signature-to-kmers.py'  is no longer there.  Is there some way I can still do this or has this been dropped? I'm using sourmash4.0 if that helps.

### LINKED ISSUE #1692: Motif predictions using protein sketch
Hello, 

Thank you so much for creating sourmash and the awesome documentation. I am interested in using sourmash for protein motif identification in PhIPseq libraries. I love the idea of identifying kmer motifs using dayhoff similarities. Antibodies likely recognized short motifs with shared chemical structure. I was wondering if there's any way to pull out the motifs from the hash table similar to issue #1372 . I tried to use the code that said it doesn't work for protein and it didn't work (haha). Even if I could pull out the dayhoff motifs that would be very interesting, but would be the coolest if I could look at the amino acid contribution to each position in the kmer. Are there any work arounds that I could get started on? I'm not the most experienced coder.

I know this may be a big feature request but I just wanted to put it out there!

### PR DESCRIPTION
**Pull request by ctb**
This PR builds on https://github.com/sourmash-bio/sourmash/pull/1653 to provide a way to get the k-mers and/or sequences underlying MinHash sketches - including DNA sketches, translated sketches, and protein/dayhoff/hp sketches.

The first addition is a new `MinHash` method, `kmers_and_hashes(seq)`, that provides matched tuples of `(kmer, hashval)` from the sequence.

The second addition is a new command-line method, `sourmash sig kmers`, that when given one or more signatures and some sequence files, will retrieve all k-mers and/or sequences that correspond to the hashes in the signatures. The usage looks like the following:

```
sourmash sig kmers --sig xxx.sig --seq podar-ref/1.fa podar-ref/2.fa \
       --save-kmers xxx.csv --save-sequences xxx.fa 
```

with the following output:
```
== This is sourmash version 4.2.1.dev9+g02f2128. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

loaded 1 signatures total, from 1 files
loaded and merged 1 signatures

merged signature has the following properties:
k=31 molecule=DNA num=0 scaled=100000 seed=42
total hashes in merged signature: 6

now processing sequence files for matches!
opening sequence file 'podar-ref/1.fa'
... searched 1486778 from 1 files so far
opening sequence file 'podar-ref/2.fa'
... searched 4150880 from 2 files so far
DONE.
searched 2 sequences from 2 files, containing a total of 4.2 Mbp.
matched and saved a total of 1 sequences with 1.5 Mbp.
matched and saved a total of 6 k-mers.
found 6 distinct matching hashes (100.0%)
```

Fixes #1372.
Fixes https://github.com/sourmash-bio/sourmash/issues/1692.
Fixes https://github.com/sourmash-bio/sourmash/issues/483.
Replaces https://github.com/sourmash-bio/sourmash/pull/724.
Ref https://github.com/sourmash-bio/sourmash/pull/477.

The code and tests are not yet complete so various edge cases may not quite work yet, but it's headed in the right direction!

TODO:
- [x] link in more issues, expand discussion
- [x] add tests for many things :)
- [x] figure out what happens to invalid DNA k-mers and suggestion resolution
- [x] evaluate and compare with functionality in https://github.com/sourmash-bio/sourmash/pull/724, and maybe pull in functionality from there
- [x] look at [this comment](https://github.com/sourmash-bio/sourmash/issues/483#issuecomment-393524023 for doc/faq information
- [x] create an issue around hash-equals-zero in Rust => None in Python (see https://github.com/sourmash-bio/sourmash/issues/1751)

### CODE REVIEW

**Comment by mr-eyes** [comment]
I was just thinking about sth similar! That's a 🆒 feature!

**Comment by ctb** [comment]
@bluegenes @glfrey @elzerac @olgabot @mr-eyes ok, I'm getting close to done with the main features, and will round out out the tests and the documentation next. At this stage I would very much appreciate feedback on the `sourmash sig kmers` output, the output formats, and anything else you are interested in seeing this code do!

Note that there is still a big potential problem with invalid DNA sequence, so, um, don't give it bad DNA sequences, k? 😆

**Comment by ctb** [comment]
ok @mr-eyes we've reached the time that I was dreading... I added some Bad DNA 🧬 into the sequences, and it all broke 😭 

https://github.com/sourmash-bio/sourmash/pull/1695/commits/b3f82e023e9ffee66dd2ac9b9417a1b2c01f9b6e

this is because bad k-mers are skipped.

what do you think about modifying your code from https://github.com/sourmash-bio/sourmash/pull/1653 to yield `None` as the hash value for k-mers in Python when 'force' is used to hash bad DNA? That seems to be fairly parsimonious to me, and shouldn't break existing code 🤞 .

**Comment by elzerac** [comment]
Hello, Thank you so much for working on this and sorry for my late reply. I had to learn for the first time how to use git to go into the developer software. Here is my implementation and it fails with the following error: 

`ERROR when merging signature 'Lyn_proteins.fasta' (b8c953b9) from file Lyn_proteins.fasta.sig
different ksizes cannot be compared`

Here is what I did to get here: 
```

#use the development version of Sourmash:

#follow the protocol here: https://github.com/sourmash-bio/sourmash/blob/add/sig_kmers/doc/developer.md#running-tests-and-checks

#first install mamba

wget https://github.com/conda-forge/miniforge/releases/latest/download/Mambaforge-$(uname)-$(uname -m).sh
bash Mambaforge-$(uname)-$(uname -m).sh

mamba create -n sourmash_dev tox-conda rust git compilers pandoc
#must re-start shell for it to work

conda activate sourmash_dev

git clone https://github.com/sourmash-bio/sourmash.git

cd sourmash

pip install -r requirements.txt

#make sure that I have the correct version from github
git branch —all

git checkout add/sig_kmers

git status

#now try to get protein signatures and use sig kmers to extract

sourmash sketch protein -p k=7,scaled=1,dayhoff -p k=8,scaled=1,dayhoff Lyn_proteins.fasta

sourmash sig kmers --sig Lyn_proteins.fasta.sig --seq Lyn_proteins.fasta --save-kmers test.csv --save-sequences test_seqs.fa
```

**Comment by ctb** [comment]
yep! you'll need to do each k-mer size individually, I'm afraid - specify `-k 8` or `-k 7`.

**Comment by elzerac** [comment]
(oh duh) sorry for spamming, but yeah the error message was very informative had I considered it for longer than a second. it was because i asked to generate kmers using two different kmer sizes. If I do the following program, it works: 

```
sourmash sketch protein -p k=7,scaled=1,dayhoff Lyn_proteins.fasta

sourmash sig kmers --sig Lyn_proteins.fasta.sig --seq Lyn_proteins.fasta --save-kmers test.csv --save-sequences test_seqs.fa
```

It's still running, but I will let you know what happens when it finishes.

**Comment by ctb** [comment]
> (oh duh) sorry for spamming, but yeah the error message was very informative

more positive feedback than this is hard to imagine :) :) (well, beyond "it actually just worked" but then usually we don't hear anything back at all 😆 )

>If I do the following program, it works:
> 
> ```
> sourmash sketch protein -p k=7,scaled=1,dayhoff Lyn_proteins.fasta
> 
> sourmash sig kmers --sig Lyn_proteins.fasta.sig --seq Lyn_proteins.fasta --save-kmers test.csv --save-sequences test_seqs.fa
> ```
> 
> It's still running, but I will let you know what happens when it finishes.

👍 

(thanks for nudging us, this was finally within reach but I hadn't realized it quite yet!)

**Comment by mr-eyes** [comment]
> what do you think about modifying your code from #1653 to yield `None` as the hash value for k-mers in Python when 'force' is used to hash bad DNA? That seems to be fairly parsimonious to me, and shouldn't break existing code 🤞 .

@ctb How returning `None` will not break existing code? I mean, if the second kmer is bad in 3 kmers-long sequence, `seq_to_hashes` should return something like [9792187361928631, None, 1290312937981723]?

**Comment by ctb** [comment]
> > what do you think about modifying your code from #1653 to yield `None` as the hash value for k-mers in Python when 'force' is used to hash bad DNA? That seems to be fairly parsimonious to me, and shouldn't break existing code 🤞 .
> 
> @ctb How returning `None` will not break existing code? I mean, if the second kmer is bad in 3 kmers-long sequence, `seq_to_hashes` should return something like [9792187361928631, None, 1290312937981723]?

Well, I mean no existing _released_ code. `seq_to_hashes` hasn't been released, I'm pretty sure, and we don't use it internally for anything (yet).

**Comment by mr-eyes** [comment]
> > > what do you think about modifying your code from #1653 to yield `None` as the hash value for k-mers in Python when 'force' is used to hash bad DNA? That seems to be fairly parsimonious to me, and shouldn't break existing code 🤞 .
> > 
> > 
> > @ctb How returning `None` will not break existing code? I mean, if the second kmer is bad in 3 kmers-long sequence, `seq_to_hashes` should return something like [9792187361928631, None, 1290312937981723]?
> 
> Well, I mean no existing _released_ code. `seq_to_hashes` hasn't been released, I'm pretty sure, and we don't use it internally for anything (yet).

Got it. Ok. In that case, can we return `0` instead of `None` for the skipped kmers while `force=True`? this is because I have to retain the type of all the values in the returned vector.

**Comment by ctb** [comment]
> > > > what do you think about modifying your code from #1653 to yield `None` as the hash value for k-mers in Python when 'force' is used to hash bad DNA? That seems to be fairly parsimonious to me, and shouldn't break existing code 🤞 .
> > > 
> > > 
> > > @ctb How returning `None` will not break existing code? I mean, if the second kmer is bad in 3 kmers-long sequence, `seq_to_hashes` should return something like [9792187361928631, None, 1290312937981723]?
> > 
> > 
> > Well, I mean no existing _released_ code. `seq_to_hashes` hasn't been released, I'm pretty sure, and we don't use it internally for anything (yet).
> 
> Got it. Ok. In that case, can we return `0` instead of `None` for the skipped kmers while `force=True`? this is because I have to retain the type of all the values in the returned vector.

My only concern with this is that `0` is a valid hash value from murmurhash, so there are edge cases where it could matter. They're very (VERY) rare tho :).

Maybe make it so the 🐍  wrapper returns None, i.e. when it sees a 0 it swaps in a None?

I'd also be slightly surprised if 🦀 didn't support some kind of enum or struct thing that would let you supply heterogenous values, but it might be too complicated or inefficient in practice.

Long term it might be nice to have `kmers_and_hashes` be implemented in 🦀 but it's not critical, just would be a nice consolidation of code.

**Comment by ctb** [comment]
thanks for #1696! May I have please?

**Comment by mr-eyes** [comment]
> My only concern with this is that `0` is a valid hash value from murmurhash, so there are edge cases where it could matter. They're very (VERY) rare tho :).

Totally agree! for easier implementation in the Rust side, I selected 0 to represent an exception as per https://stackoverflow.com/a/54695423/3371177 , however, this should definitely be replaced in the future.
 
> Maybe make it so the 🐍 wrapper returns None, i.e. when it sees a 0 it swaps in a None?
>
> I'd also be slightly surprised if 🦀 didn't support some kind of enum or struct thing that would let you supply heterogenous values, but it might be too complicated or inefficient in practice.
>
> Long term it might be nice to have `kmers_and_hashes` be implemented in 🦀 but it's not critical, just would be a nice consolidation of code.

Well, I have already started implementing it in Rust https://github.com/sourmash-bio/sourmash/tree/mo/rust_seq_and_kmers and will probably open a PR soon :)

**Comment by mr-eyes** [comment]
> thanks for #1696! May I have please?

Yes, please! :)

**Comment by ctb** [comment]
With the merge of #1696 and the switch to `kmers_and_hashes` providing `None` where k-mers are Bad, I think we're close to having this be fully baked. I want to add some tests for bad k-mers in `sourmash sig kmers`, and of course want to add some docs. Other thoughts and suggestions welcome, but they'll have to come in the next few days to make it into this PR :)

**Comment by ctb** [comment]
I think this is ready for review @sourmash-bio/devs!

**Review by mr-eyes — COMMENTED** [review]
I was testing the `sig kmers` cli and have some comments.
All files used in the testing is attached in the zip file.

[sig_kmers_mo_test_files.zip](https://github.com/sourmash-bio/sourmash/files/6926663/sig_kmers_mo_test_files.zip)

### Creating the sketch
```
sourmash sketch dna ref.fa -o ref.fa.sig -p k=31,scaled=1 --check-sequence
```

### Perform sig kmers on a typical sequence to the ref.fa

```
sourmash sig kmers --signatures ref.fa.sig --sequences query_1.fa --save-sequences matches_1.fasta --save-kmers kmer-matches_1.csv --check-sequence
```
```
loaded 1 signatures total, from 1 files
loaded and merged 1 signatures

merged signature has the following properties:
k=31 molecule=DNA num=0 scaled=1 seed=42
total hashes in merged signature: 35

now processing sequence files for matches!
opening sequence file 'query_1.fa'
DONE.
searched 1 sequences from 1 files, containing a total of 0.0 Mbp.
matched and saved a total of 1 sequences with 0.0 Mbp.
matched and saved a total of 35 k-mers.
found 35 distinct matching hashes (100.0%)
```
---
### Perform sig kmers on a typical sequence to the ref.fa **with an extra valid kmer**
```
sourmash sig kmers --signatures ref.fa.sig --sequences query_3.fa --save-sequences matches_3.fasta --save-kmers kmer-matches_3.csv --check-sequence
```
```
loaded 1 signatures total, from 1 files
loaded and merged 1 signatures

merged signature has the following properties:
k=31 molecule=DNA num=0 scaled=1 seed=42
total hashes in merged signature: 35

now processing sequence files for matches!
opening sequence file 'query_3.fa'
DONE.
searched 1 sequences from 1 files, containing a total of 0.0 Mbp.
matched and saved a total of 1 sequences with 0.0 Mbp.
matched and saved a total of 35 k-mers.
found 36 distinct matching hashes (100.0%)
```

**Q1. I think there are only 35 distinct matching hashes, not 36. Am I getting it right?**

---

---
### Perform sig kmers on a typical sequence to the ref.fa **with an extra (bad) kmer**
```
sourmash sig kmers --signatures ref.fa.sig --sequences query_2.fa --save-sequences matches_2.fasta --save-kmers kmer-matches_2.csv --check-sequence --force
```
```
loaded 1 signatures total, from 1 files
loaded and merged 1 signatures

merged signature has the following properties:
k=31 molecule=DNA num=0 scaled=1 seed=42
total hashes in merged signature: 35

now processing sequence files for matches!
opening sequence file 'query_2.fa'
ERROR in sequence 'with_badkmer_at_the_end', file 'query_2.fa'
invalid DNA character in input k-mer: GCATCGACTAGCTACGGCGATCGACTAAACN
(continuing)
DONE.
searched 0 sequences from 1 files, containing a total of 0.0 Mbp.
matched and saved a total of 0 sequences with 0.0 Mbp.
matched and saved a total of 0 k-mers.
found 0 distinct matching hashes (0.0%)
```

**Comment: I think the last extra bad kmer should be skipped and show that there's another 35 matched hashes, so the result files in this step should match the query_1 output files.**

**Comment by ctb** [comment]
Thanks @mr-eyes - fixed, with extra tests now! 😎

>Q1. I think there are only 35 distinct matching hashes, not 36. Am I getting it right?

You are correct. I was counting all of the hashes in the sequence, not just the ones in the query sig!

>Comment: I think the last extra bad kmer should be skipped and show that there's another 35 matched hashes, so the result files in this step should match the query_1 output files.

The `MinHash.add_sequence(...)` raises an exception if there's a bad character in the _sequence itself_, so we don't have the granularity to add the individual k-mers at this time.

Ready for re-review :)

**Comment by mr-eyes** [comment]
Thank you!

> The `MinHash.add_sequence(...)` raises an exception if there's a bad character in the _sequence itself_, so we don't have the granularity to add the individual k-mers at this time.

So, `--force` should skip the whole sequence instead of just the bad kmers in it and continue querying the rest?

**Comment by ctb** [comment]
> Thank you!
> 
> > The `MinHash.add_sequence(...)` raises an exception if there's a bad character in the _sequence itself_, so we don't have the granularity to add the individual k-mers at this time.
> 
> So, `--force` should skip the whole sequence instead of just the bad kmers in it and continue querying the rest?

yes, that's what `--force` with `--check-sequence` **does**. Whether it *should* is maybe a different question 😄 

Note that without `--check-sequence`, the sequence is processed normally.

Incidentally, I did update the error message to make it clear it was skipping the entire sequence.

File: src/sourmash/sig/__main__.py [hunk 1]
```diff
@@ -916,6 +917,208 @@ def export(args):
...
+    if not n_sequences_searched:
+        notify("ERROR: no sequences searched!?")
+        sys.exit(-1)
+
+    # ...and report!
+    notify("DONE.")
+    notify(f"searched {n_sequences_searched} sequences from {n_files_searched} files, containing a total of {n_bp_searched/1e6:.1f} Mbp.")
+
+    if save_seqs:
+        notify(f"matched and saved a total of {n_sequences_found} sequences with {n_bp_saved/1e6:.1f} Mbp.")
```
Comment (mr-eyes) [inline src/sourmash/sig/__main__.py]:
```suggestion
    # ...and report!
    def len_label(no_bp):
        n = 0
        formatFloat = lambda n: f"{n:.1f}" if n%1 else int(n)
        mag_labels = {0 : '', 1: 'K', 2: 'M', 3: 'G', 4: 'T'}
        while no_bp > 2**10: no_bp /= 2**10; n += 1
        return f"{formatFloat(no_bp)} {mag_labels[n]}bp"
        
    notify("DONE.")
    notify(f"searched {n_sequences_searched} sequences from {n_files_searched} files, containing a total of {len_label(n_bp_searched)}.")

    if save_seqs:
        notify(f"matched and saved a total of {n_sequences_found} sequences with {len_label(n_bp_saved)}.")
```

Comment (mr-eyes) [reply src/sourmash/sig/__main__.py]: This is not a really important suggestion, but it will give a more flexible label.

**Review by mr-eyes — APPROVED** [review]
Looks great, thank you!

File: src/sourmash/sig/__main__.py (hunk 1 above)
Comment (ctb) [reply src/sourmash/sig/__main__.py]: fine, you guilted me into it! See 2b4d7d31 and 0f0b0d016f4 😆

**Comment by ctb** [comment]
will merge when tests pass.

File: doc/api-example.md [hunk 2]
```diff
@@ -119,15 +131,15 @@ True
...
-of 42 and taking the first 64 bits.
+chosen for hashing by sourmash. This method was chosen to be compatible
+with [mash](https://mash.readthedocs.io/.
 
 ### Protein-based encodings
 
-By default, `MinHash` objects work with DNA. However, sourmash supports
-amino acid, Dayhoff, and hydrophobic-polar encodings as well.
+By default, `MinHash` objects work with DNA. However, sourmash
+supports amino acid, Dayhoff, and hydrophobic-polar 'hp) encodings as
```
Comment (luizirber) [inline doc/api-example.md]:
```suggestion
supports amino acid, Dayhoff, and hydrophobic-polar (hp) encodings as
```

**Comment by ctb** [comment]
hah! I thought I'd merged this already and was surprised to receive a suggestion from @luizirber 😆 . Thanks for the reminder (and the fix!)

**Comment by ctb** [comment]
🎉
